experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+
+# This file contains the relational SSA dialect, a dialect that expresses
+# relational queries using SSA. The dialect consists of two types of operations:
+# Expressions and Operators. Operators work on Bags whereas expressions work on
+# single values and are used to specify the exact behavior of operators. The
+# Column and Yield Expressions bridge the gap to convert from one to the other.
+
+#===------------------------------------------------------------------------===#
+# Data types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational SSA query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational SSA query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational SSA query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+#===------------------------------------------------------------------------===#
+# Query types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_ssa.bool
+  '''
+  """
+  name = "rel_ssa.bool"
+
+
+@irdl_attr_definition
+class SchemaElement(ParametrizedAttribute):
+  """
+  Models an element of a schema with name `elt_name` and type `elt_type`.
+
+  Example:
+  '''
+  !rel_ssa.schema_element<"id", !rel_ssa.int32>
+  '''
+  """
+  name = "rel_ssa.schema_element"
+
+  elt_name = ParameterDef(StringAttr)
+  elt_type = ParameterDef(DataType)
+
+  @staticmethod
+  @builder
+  def get(name: str, type_: DataType) -> 'SchemaElement':
+    return SchemaElement([StringAttr.from_str(name), type_])
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
+  """
+  Models a bag in a relational SSA query. The exact schema of the bag is part of
+  the type itself.
+
+  Example:
+
+  '''
+  !rel_ssa.bag<[!rel_ssa.schema_element<"id", !rel_ssa.int32>]>
+  '''
+  """
+  name = "rel_ssa.bag"
+
+  schema = ParameterDef(ArrayOfConstraint(SchemaElement))
+
+  @staticmethod
+  @builder
+  def get(types: list[DataType], names: list[str]) -> 'Bag':
```

Comment:
nit: your logic, printing and parsing use `(name, type)` but your API uses `(type, name)`.
Can you make the API follow the rest for consistency?